src/v/cluster_link/replication/mux_remote_consumer.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cluster_link/replication/partition_data_queue.h"
+#include "container/chunked_hash_map.h"
+#include "kafka/client/direct_consumer/direct_consumer.h"
+
+#include <memory>
+
+class MuxConsumerFixture;
+
+namespace cluster_link::replication {
+
+/**
+ * @brief Multiplexed remote consumer for managing multiple partition consumers
+ *
+ * This class provides a multiplexed consumer that can manage multiple
+ * partitions across different topics using a single underlying Kafka consumer.
+ * It handles fetching data from remote partitions and distributing it to
+ * individual partition data queues. This is done so that an individual slow
+ * partition does not block the fetching of data from remaining partitions.
+ *
+ * The consumer operates with a fetch loop that retrieves data from all assigned
+ * partitions and buffers it according to the configured limits. Partitions can
+ * be dynamically added or removed during runtime depending on if their queue is
+ * full.
+ */
+class mux_remote_consumer {
+public:
+    explicit mux_remote_consumer(
+      std::unique_ptr<kafka::client::direct_consumer> consumer,
+      size_t partition_max_buffered,
+      std::chrono::milliseconds fetch_max_wait)
+      : _consumer(std::move(consumer))
+      , _partition_max_buffered(partition_max_buffered)
+      , _fetch_max_wait(fetch_max_wait) {}
+
+    ss::future<> start();
+    ss::future<> stop() noexcept;
+
+    void add(const ::model::topic_partition&);
+    ss::future<> remove(const ::model::topic_partition&);
+
+    /**
+     * Reset the state of a partition to the provided offset.
+     * Aborts any inflight fetch with abort_requested_exception.
+     * Any subsequent fetch will start from the new offset.
+     *
+     * Also used to set the seed offset of fetches after add().
+     */
+    void reset(const ::model::topic_partition& tp, kafka::offset);
+
+    /**
+     * Fetch data for the specified partition. Pre-reqs for a fetch
+     * are add() and reset() (atleast once). add() adds the partition to the
+     * consumer and reset() sets the initial offset for fetching.
+     */
+    ss::future<partition_data_queue::fetch_data>
```

Comment:
this part seems to be leaking units all the way from `partition_data_queue`, is it possible the queue is destroyed before batches are consumed ? 
If we need to manage total memory usage maybe it would be better to have the per shard sidecar component that would account for all inflight batches memory usage ?